### PR TITLE
Add option to start stream from a specific version

### PR DESCRIPTION
**Pull request by SpaceRangerWes**
**Goal:** Add the ability to start structured streaming from a user provided version as outlined in #474 

**Desc:** This change will introduce `startingVersion` and `startingTimestamp` to the Delta Stream Reader so that users can define a starting point of their stream source without processing the entire table.

**What's the behavior if both startingVersion and startingTimestamp are set?** Both options cannot be provided together. An error will be thrown telling you to choose only one.

**When a query restarts, how does the new options work?** If a query restarts and a checkpoint location is provided, the stream will start from the checkpoints offset as opposed to the provided option.

**If the user uses an old Delta Lake version to read the checkpoint, what will happen?** If we use a newer version of Delta with this option, we will store an offset that looks like: Process up to this version and this index. If the stream fails and someone reverts to an older version of Delta, we would try to process ALL the data in the snapshot up to the end offset which we came up with.

Credit to @brkyvz for a large percentage of the work.

### CODE REVIEW

File: src/main/scala/org/apache/spark/sql/delta/sources/DeltaSource.scala [hunk 1]
```diff
@@ -102,6 +104,9 @@ case class DeltaSource(
   private var initialState: DeltaSourceSnapshot = null
   private var initialStateVersion: Long = -1L
 
+  final val STARTING_VERSION_KEY = "startingVersion"
```
Comment (brkyvz) [inline src/main/scala/org/apache/spark/sql/delta/sources/DeltaSource.scala:107]: can you put these into `DeltaOptions`?

File: src/main/scala/org/apache/spark/sql/delta/sources/DeltaSource.scala [hunk 2]
```diff
@@ -371,4 +381,30 @@ case class DeltaSource(
...
+
+  /** Extracts whether users provided the option to time travel a relation. */
+  private def getTimeTravelVersion(
+    parameters: CaseInsensitiveMap[String]): Option[(Long, String)] = {
+    val tsOpt = parameters.get(STARTING_TIMESTAMP_KEY)
+    val versionOpt = parameters.get(STARTING_VERSION_KEY)
+
+    if (tsOpt.isDefined && versionOpt.isDefined) {
+      throw DeltaErrors.provideOneOfInTimeTravel // TODO: change error
+    } else if (tsOpt.isDefined) {
```
Comment (brkyvz) [inline src/main/scala/org/apache/spark/sql/delta/sources/DeltaSource.scala:393]: this logic can be in DeltaOptions

File: src/main/scala/org/apache/spark/sql/delta/sources/DeltaSource.scala [hunk 3]
```diff
@@ -371,4 +381,30 @@ case class DeltaSource(
       case other => throw new UnsupportedOperationException(s"Unknown ReadLimit: $other")
     }
   }
+
+  /** Extracts whether users provided the option to time travel a relation. */
+  private def getTimeTravelVersion(
```
Comment (brkyvz) [inline src/main/scala/org/apache/spark/sql/delta/sources/DeltaSource.scala:386]: It's not really time travel in the case for streaming. Thinking about it more, all of this logic can live within DeltaOptions as well

Comment (SpaceRangerWes) [reply src/main/scala/org/apache/spark/sql/delta/sources/DeltaSource.scala:386]: Yep. Stole this from a different module. Just haven't retitled it. I'll refactor it to DeltaOptions for both impacted modules https://github.com/delta-io/delta/blob/e365c2d3db3bf5bd9c70888fd704f43c98393837/src/main/scala/org/apache/spark/sql/delta/sources/DeltaDataSource.scala#L324

**Comment by SpaceRangerWes** [comment]
@brkyvz I didn't move all of the `getStartingVersion` logic to `DeltaOptions` because it seemed too far outside the responsibility/scope of an options module to resolve the time travel version. However, the input specific validation has been moved.

Should we add a test that somehow validates that prior versions weren't considered? If so I'll need some guidance on how to get that info in the unit test.

File: src/main/scala/org/apache/spark/sql/delta/DeltaOptions.scala [hunk 4]
```diff
@@ -121,6 +121,25 @@ trait DeltaReadOptions extends DeltaOptionParser {
       throw new IllegalArgumentException(
         s"Please recheck your syntax for '$EXCLUDE_REGEX_OPTION'", e)
   }
+
+  val startingVersion = options.get(STARTING_VERSION_OPTION).map{ str =>
```
Comment (brkyvz) [inline src/main/scala/org/apache/spark/sql/delta/DeltaOptions.scala:125]: nit: space after `map`

File: src/test/scala/org/apache/spark/sql/delta/DeltaSourceSuite.scala [hunk 5]
```diff
@@ -991,6 +993,146 @@ class DeltaSourceSuite extends DeltaSourceSuiteBase {
...
+  test("startingTimestamp: invalid inputs") {
+    withTempDirs { (inputDir, outputDir, checkpointDir) =>
+      val tblLoc = inputDir.getCanonicalPath
+      val start = 1594418440842L
+      generateCommits(tblLoc, start, start + 20.minutes)
+
+      val e = intercept[StreamingQueryException] {
+        spark.readStream
+          .format("delta")
+          .option("startingTimestamp", "2019-07-10 15:20:40")
```
Comment (brkyvz) [inline src/test/scala/org/apache/spark/sql/delta/DeltaSourceSuite.scala:1106]: what error do we throw if we give a bad timestamp? Also can we also please test `"2019-07-10"`

File: src/main/scala/org/apache/spark/sql/delta/DeltaOptions.scala [hunk 6]
```diff
@@ -121,6 +122,37 @@ trait DeltaReadOptions extends DeltaOptionParser {
...
+
+  val startingVersion = options.get(STARTING_VERSION_OPTION).map { str =>
+    Try(str.toLong).toOption.filter(_ >= 0).getOrElse{
+      throw DeltaErrors.illegalDeltaOptionException(
+        STARTING_VERSION_OPTION, str, "must be greater than or equal to zero")
+    }
+  }
+
+  val startingTimestamp = options.get(STARTING_TIMESTAMP_OPTION).map { str =>
+    val pattern = "yyyy-MM-dd HH:mm:ss"
```
Comment (brkyvz) [inline src/main/scala/org/apache/spark/sql/delta/DeltaOptions.scala:134]: oh, we don't need to add any specific pattern check here. I was hoping we can leave it to Spark to figure it out

File: src/main/scala/org/apache/spark/sql/delta/sources/DeltaSource.scala [hunk 7]
```diff
@@ -371,4 +376,19 @@ case class DeltaSource(
...
+  /** Extracts whether users provided the option to time travel a relation. */
+  private def getStartingVersion: Option[(Long, String)] = {
+    val tsOpt = options.startingTimestamp
+    val versionOpt = options.startingVersion
+
+    /** DeltaOption validates input and ensures that only one is provided. */
+    if (tsOpt.isDefined || versionOpt.isDefined) {
+      Some(DeltaTableUtils.resolveTimeTravelVersion(
+        spark.sessionState.conf, deltaLog,
+        DeltaTimeTravelSpec(tsOpt.map(Literal(_)), versionOpt, Some("streamSource"))))
```
Comment (brkyvz) [inline src/main/scala/org/apache/spark/sql/delta/sources/DeltaSource.scala:389]: my bad: sorry, changed my mind here, can we change `streamSource` to `deltaSource` please

File: src/test/scala/org/apache/spark/sql/delta/DeltaSourceSuite.scala [hunk 8]
```diff
@@ -991,6 +995,162 @@ class DeltaSourceSuite extends DeltaSourceSuiteBase {
...
+        .load(inputDir.getCanonicalPath)
+        .writeStream
+        .format("delta")
+        .option("checkpointLocation", checkpointDir.getCanonicalPath)
+        .start(outputDir.getCanonicalPath)
+      try {
+        q.processAllAvailable()
+        checkAnswer(
+          spark.read.format("delta").load(outputDir.getAbsolutePath),
+          (0 until 20).map(_.toLong).toDF())
```
Comment (brkyvz) [inline src/test/scala/org/apache/spark/sql/delta/DeltaSourceSuite.scala:1035]: why is this not 10 until 20?

File: src/test/scala/org/apache/spark/sql/delta/DeltaSourceSuite.scala [hunk 9]
```diff
@@ -991,6 +995,162 @@ class DeltaSourceSuite extends DeltaSourceSuiteBase {
...
+      }
+
+      assert(e2.getCause.isInstanceOf[AnalysisException])
+      assert(e2.getCause.getMessage.contains("The provided timestamp: 2020-07-10 17:20:40.0" +
+        " is after the latest commit timestamp"))
+
+      val e3 = intercept[StreamingQueryException] {
+        spark.readStream
+          .format("delta")
+          .option("startingTimestamp", "2020-07-10")
```
Comment (brkyvz) [inline src/test/scala/org/apache/spark/sql/delta/DeltaSourceSuite.scala:1140]: This should work, it shouldn't throw an exception

**Comment by zsxwing** [comment]
@SpaceRangerWes Thanks a lot for your PR. Since this one is adding new user facing features, could you update the PR description to describe what are the new options? It would be also great if you can answer the following questions in the PR description:
- What's the behavior if both `startingVersion` and startingTimestamp are set?
- When a query restarts, how does the new options work?
- If the user uses an old Delta Lake version to read the checkpoint, what will happen?

File: src/test/scala/org/apache/spark/sql/delta/DeltaSourceSuite.scala (hunk 8 above)
Comment (SpaceRangerWes) [reply src/test/scala/org/apache/spark/sql/delta/DeltaSourceSuite.scala:1035]: Just because it is a later version, wouldn't that version also contain all previously written data from version 0?

Comment (brkyvz) [reply src/test/scala/org/apache/spark/sql/delta/DeltaSourceSuite.scala:1035]: we want to pull the incremental changes made to the table from that version onwards, so I'd expect this to start at 10, not include 0-10

File: src/main/scala/org/apache/spark/sql/delta/sources/DeltaSource.scala [hunk 10]
```diff
@@ -177,12 +177,19 @@ case class DeltaSource(
...
+    val (version, isStartingVersion) = getStartingVersion match {
+      case Some(v) => (v._1, false)
+      case None => (deltaLog.snapshot.version, true)
+    }
     if (version < 0) {
       return None
     }
+    if (options.startingVersion.isDefined || options.startingTimestamp.isDefined) {
+      return Some(DeltaSourceOffset(tableId, version, -1, isStartingVersion = false))
+    }
```
Comment (brkyvz) [inline src/main/scala/org/apache/spark/sql/delta/sources/DeltaSource.scala:190]: I don't think this is the right thing to return here. `getStartingOffset` returns the "endOffset" for the first micro-batch. What you were doing in this method before was right

File: src/main/scala/org/apache/spark/sql/delta/sources/DeltaSource.scala [hunk 11]
```diff
@@ -371,4 +380,19 @@ case class DeltaSource(
       case other => throw new UnsupportedOperationException(s"Unknown ReadLimit: $other")
     }
   }
+
+  /** Extracts whether users provided the option to time travel a relation. */
+  private def getStartingVersion: Option[(Long, String)] = {
```
Comment (brkyvz) [inline src/main/scala/org/apache/spark/sql/delta/sources/DeltaSource.scala:385]: Can this method just return the version instead of the tuple?

File: src/main/scala/org/apache/spark/sql/delta/sources/DeltaSource.scala [hunk 12]
```diff
@@ -285,6 +292,8 @@ case class DeltaSource(
     val changes = if (start.isEmpty) {
       if (endOffset.isStartingVersion) {
```
Comment (brkyvz) [inline src/main/scala/org/apache/spark/sql/delta/sources/DeltaSource.scala:293]:
```scala
getStartingVersion match {
  case Some(v) => getChanges(v, -1L, isStartingVersion = false)
  case _ =>
     if (endOffset.isStartingVersion) {
       getChanges(endOffset.reservoirVersion, -1L, isStartingVersion = true)
     } else {
        assert(endOffset.reservoirVersion > 0, s"invalid reservoirVersion in endOffset: $endOffset")
        // Load from snapshot `endOffset.reservoirVersion - 1L` so that `index` in `endOffset`
        // is still valid.
        getChanges(endOffset.reservoirVersion - 1L, -1L, isStartingVersion = true)
      }
}

```

File: src/main/scala/org/apache/spark/sql/delta/sources/DeltaSource.scala (hunk 11 above)
Comment (brkyvz) [reply src/main/scala/org/apache/spark/sql/delta/sources/DeltaSource.scala:385]: Also, can it be a `lazy val`, because time travel resolution for timestamp can take a while and we may re-use this value

File: src/main/scala/org/apache/spark/sql/delta/sources/DeltaSource.scala [hunk 13]
```diff
@@ -285,6 +292,8 @@ case class DeltaSource(
...
         getChanges(endOffset.reservoirVersion, -1L, isStartingVersion = true)
+      } else if (options.startingVersion.isDefined || options.startingTimestamp.isDefined) {
+        getChanges(endOffset.reservoirVersion, -1L, isStartingVersion = false)
       } else {
         assert(endOffset.reservoirVersion > 0, s"invalid reservoirVersion in endOffset: $endOffset")
```
Comment (zsxwing) [inline src/main/scala/org/apache/spark/sql/delta/sources/DeltaSource.scala:298]: IIUC, we are introducing a new call to getBatch like this (start is `None` and `isStartingVersion` in end is `false`):

```
getBatch(
  start = None,
  end = `{"sourceVersion":1,
          "reservoirId":"testId",
          "reservoirVersion":1,
          "index":-1,
          "isStartingVersion":false}`
  )
```

- If `reservoirVersion` is 0 (using `startingVersion = 0`), then the query running without this patch will fail.
- If `reservoirVersion` is greater than 0, then the query running without this patch will return a different answer: it returns all files in `getSnapshot(reservoirVersion)`.

Of cause, this is an issue only in the first batch and only when
- The offset log of the first batch is generated in a new Delta Lake version.
- The first batch fails.
- Rollback to the old version and rerun it. In this case, the  re-generated batch will be different than the previous run.

I think this should be okay. It's really hard to hit this issue. @marmbrus what do you think?

File: src/main/scala/org/apache/spark/sql/delta/sources/DeltaSource.scala (hunk 12 above)
Comment (zsxwing) [reply src/main/scala/org/apache/spark/sql/delta/sources/DeltaSource.scala:293]: I think we need to use `start` and `start.get.isStartingVersion` passed into `getBatch`.

File: src/main/scala/org/apache/spark/sql/delta/sources/DeltaSource.scala (hunk 10 above)
Comment (SpaceRangerWes) [reply src/main/scala/org/apache/spark/sql/delta/sources/DeltaSource.scala:190]: @brkyvz when I do that, the test _"startingVersion: start from version > 0"_ doesn't return a result. The wrong offset is returned.

Comment (brkyvz) [reply src/main/scala/org/apache/spark/sql/delta/sources/DeltaSource.scala:190]: The idea is that this method starts from startingVersion index -1, and finds the last available element after that, which should be startingVersion + 1 index -1

File: src/main/scala/org/apache/spark/sql/delta/sources/DeltaSource.scala (hunk 11 above)
Comment (SpaceRangerWes) [reply src/main/scala/org/apache/spark/sql/delta/sources/DeltaSource.scala:385]: Wouldn't we want `final private def getStartingVersion` if we only want it to initialize only once?

Comment (brkyvz) [reply src/main/scala/org/apache/spark/sql/delta/sources/DeltaSource.scala:385]: no, final def would still evaluate the method multiple times, but prevents it from being overridden. In Scala `lazy val` is used to evaluate once and store the result only when called

File: src/main/scala/org/apache/spark/sql/delta/sources/DeltaSource.scala (hunk 13 above)
Comment (brkyvz) [reply src/main/scala/org/apache/spark/sql/delta/sources/DeltaSource.scala:298]: I also do think it is really hard. I think we should document it, but I don't think it warrants a protocol change

File: src/main/scala/org/apache/spark/sql/delta/sources/DeltaSource.scala (hunk 11 above)
Comment (SpaceRangerWes) [reply src/main/scala/org/apache/spark/sql/delta/sources/DeltaSource.scala:385]: So are you looking for something like this?

```
lazy val result = if (tsOpt.isDefined || versionOpt.isDefined) {
      Some(DeltaTableUtils.resolveTimeTravelVersion(
        spark.sessionState.conf, deltaLog,
        DeltaTimeTravelSpec(tsOpt.map(Literal(_)), versionOpt, Some("deltaSource")))._1)
    } else {
      None
    }
    
    result
```

I thought normally you assign the result of a function to a lazy val e.g. `lazy val X = getStartingVersion`

File: src/main/scala/org/apache/spark/sql/delta/sources/DeltaSource.scala (hunk 10 above)
Comment (SpaceRangerWes) [reply src/main/scala/org/apache/spark/sql/delta/sources/DeltaSource.scala:190]: If that were the case, the user would provide `startingVersion = 0` if they wanted to start at version 1... right?

File: src/main/scala/org/apache/spark/sql/delta/sources/DeltaSource.scala (hunk 11 above)
Comment (brkyvz) [reply src/main/scala/org/apache/spark/sql/delta/sources/DeltaSource.scala:385]: you could keep it a function and then assign it to a lazy val as well. Maybe have a ```lazy val startingVersionOpt = getStartingVersion```

File: src/main/scala/org/apache/spark/sql/delta/sources/DeltaSource.scala (hunk 12 above)
Comment (brkyvz) [reply src/main/scala/org/apache/spark/sql/delta/sources/DeltaSource.scala:293]: this is the case for `start` being empty though

**Comment by SpaceRangerWes** [comment]
> We're close to the finish line! Can we also add a couple more tests:
> 
> 1. Usage of startTimestamp and startVersion should throw an error
> 2. Start timestamp as a date.

@brkyvz  for (1.) options that aren't valid go through `verifyOptions()` and then log a `delta.option.invalid` message. Do you want to follow a different pattern for this?

File: src/main/scala/org/apache/spark/sql/delta/sources/DeltaSource.scala (hunk 10 above)
Comment (SpaceRangerWes) [reply src/main/scala/org/apache/spark/sql/delta/sources/DeltaSource.scala]: Fixed after I used your code snippet from getBatch

**Comment by mukulmurthy** [comment]
Sorry if this was discussed before, but this is a little confusing to me:

> When a query restarts, how does the new options work? If a query restarts and a checkpoint location is provided, the stream will start from the checkpoints offset as opposed to the provided option.

Why silently drop the option instead of throwing?

**Comment by brkyvz** [comment]
@mukulmurthy That's how all streams work. If a stream has checkpoints, they continue from the offsets persisted in the checkpoint. Some options are just provided as starting conditions. Imagine your stream dies and restarts, it would never be able to restart if we throw an error. If they need to stop their stream after the first batch, remove the option and keep going, that's horrible UX

**Comment by mukulmurthy** [comment]
Makes sense, thanks for the explanation

File: src/test/scala/org/apache/spark/sql/delta/DeltaSourceSuite.scala [hunk 14]
```diff
@@ -21,8 +21,9 @@ import java.net.URI
 import java.util.UUID
 
 import org.apache.spark.sql.delta.actions.{AddFile, InvalidProtocolVersionException, Protocol}
-import org.apache.spark.sql.delta.sources.{DeltaSourceOffset, DeltaSQLConf}
+import org.apache.spark.sql.delta.sources.{DeltaSQLConf, DeltaSourceOffset}
```
Comment (brkyvz) [inline src/test/scala/org/apache/spark/sql/delta/DeltaSourceSuite.scala:24]: can you revert this change please

File: src/main/scala/org/apache/spark/sql/delta/sources/DeltaSource.scala [hunk 15]
```diff
@@ -371,4 +380,21 @@ case class DeltaSource(
       case other => throw new UnsupportedOperationException(s"Unknown ReadLimit: $other")
     }
   }
+
+  /** Extracts whether users provided the option to time travel a relation. */
+  private def getStartingVersion: Option[Long] = {
```
Comment (brkyvz) [inline src/main/scala/org/apache/spark/sql/delta/sources/DeltaSource.scala:385]: nit: can we either cache the result of this as a lazy val or just make this method a lazy val? The computation may be a bit expensive
